### PR TITLE
check every comment token before printing it

### LINKED ISSUE #396: duplicated comment lines
**Terser version**: v4.1.2

**src**: `input.js`

``` javascript
var foo = {};
// this is a comment line
(foo.bar = window.WeakSet).foo_bar = 123
```

**command line**: `./node_modules/.bin/terser --comments=all input.js`

**real output**:

``` javascript
var foo={};
// this is a comment line
// this is a comment line
(foo.bar=window.WeakSet).foo_bar=123;
```

**expected output**:

``` javascript
var foo={};
// this is a comment line
(foo.bar=window.WeakSet).foo_bar=123;
```

**debugging**:

I find this bug is only triggered when the following line starts with a '`(`'.

### PR DESCRIPTION
**Pull request by gdh1995**
This is a simple solution to fix #396, and has very little negative impact on speed.

### CODE REVIEW

File: lib/output.js [hunk 1]
```diff
@@ -632,10 +633,11 @@ function OutputStream(options) {
             }
         }
 
-        comments = comments.filter(comment_filter, node);
+        comments = comments.filter(comment_filter, node).filter(function(c) { return c._dumped !== self; });
```
Comment (L2jLiga) [inline lib/output.js]: I would like to use arrow function here

File: lib/output.js [hunk 2]
```diff
@@ -622,7 +622,8 @@ function OutputStream(options) {
         }
 
         if (current_pos == 0) {
-            if (comments.length > 0 && options.shebang && comments[0].type == "comment5") {
+            if (comments.length > 0 && options.shebang && comments[0].type == "comment5"
```
Comment (L2jLiga) [inline lib/output.js]: May be strict equality?
`comments[0].type === "comment5"`

File: lib/output.js (hunk 1 above)
Comment (gdh1995) [reply lib/output.js]: OK. I once thought this project is using ES5 code only. Then I'll update these lines.

File: test/mocha/comments.js [hunk 3]
```diff
@@ -260,6 +260,34 @@ describe("comments", function() {
         ].join("\n"));
     });
 
+    it("Should not duplicate a same comment", function() {
+        var result = UglifyJS.minify([
+            "var foo = {};",
```
Comment (L2jLiga) [inline test/mocha/comments.js]: Let use [`template strings`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Template_literals)

File: lib/output.js (hunk 2 above)
Comment (gdh1995) [reply lib/output.js]: I think most old code is using "`.type == ""`"...

File: lib/output.js (hunk 1 above)
Comment (L2jLiga) [reply lib/output.js]: Terser uses ES6+ features, like arrow fns, Map, Set, literal strings and so on

File: lib/output.js (hunk 2 above)
Comment (L2jLiga) [reply lib/output.js]: Yes, it is. We can change this step-by-step. I mean not only `==` vs `===` but the whole code, like replacing `Object.create(null)` with `new Map()` or `new Set()`

File: test/mocha/comments.js [hunk 4]
```diff
@@ -260,6 +260,30 @@ describe("comments", function() {
...
+/* this is a block line */
+(foo.bar = {}).test = 123;`, {
+            output: { comments: true }
+        });
+        if (result.error) throw result.error;
+        assert.strictEqual(result.code, 
+`var foo={};
+/* this is a block line */(foo.bar={}).test=123;`);
+    });
+
```
Comment (fabiosantoscode) [inline test/mocha/comments.js]: I'm not happy to make this a mocha test. Terser output can change at will.

Can you make this a compress test? An example of a compress test ensuring that comments are as expected is this one: https://github.com/terser-js/terser/blob/master/test/compress/pure_funcs.js#L297-L313

**Comment by fabiosantoscode** [comment]
Thank you very much for this changeset :) I only have a nitpick about the testing, as mocha tests asserting for an exact output are very fragile.

On another note, can a WeakSet be used to record every comment that the Output instance has yielded up until this point? Something like `this.printed_comments = new WeakSet()`. These structures are fast and very cheap in terms of memory.

**Comment by fabiosantoscode** [comment]
Nevermind the set, I just checked the implementation and `_dumped` is already a thing, feel free to use it. I'll see what I can do to remove it though!

In regards to tests, feel free to create a new file `test/compress/comments.js` and place that there. I'll have to put more comments tests there since there are some issues with the license comments.

**Comment by fabiosantoscode** [comment]
Thanks @gdh1995, awesome work :D

**Pull request merged**